shell/platform/windows/flutter_windows_engine.cc

```diff
@@ -240,13 +244,109 @@ bool FlutterWindowsEngine::Run() {
   return Run("");
 }
 
+int GetReleaseVersionAndBuildNumber(ReleaseVersion* release_version) {
+  char module_path[MAX_PATH];
+  // Get the full path of the currently running executable
+  if (GetModuleFileNameA(NULL, module_path, MAX_PATH) == -1) {
+    return -1;
+  }
+
+  // Get the size of the version information
+  DWORD handle = -1;
+  DWORD version_info_size = GetFileVersionInfoSizeA(module_path, &handle);
+  if (version_info_size == -1) {
+    return -1;
+  }
+
+  // Allocate memory for version info
+  std::vector<char> version_data(version_info_size);
+  if (!GetFileVersionInfoA(module_path, handle, version_info_size,
+                           version_data.data())) {
+    return -1;
+  }
+
+  // Get the version info structure
+  VS_FIXEDFILEINFO* file_info = nullptr;
+  UINT file_info_size = -1;
+  if (!VerQueryValueA(version_data.data(), "\\",
+                      reinterpret_cast<LPVOID*>(&file_info), &file_info_size)) {
+    return -1;
+  }
+
+  if (file_info) {
+    // Extract version numbers
+    DWORD major = HIWORD(file_info->dwFileVersionMS);
+    DWORD minor = LOWORD(file_info->dwFileVersionMS);
+    DWORD build = HIWORD(file_info->dwFileVersionLS);
+
+    char version[49];
+    snprintf(version, sizeof(version), "%lu.%lu.%lu", major, minor, build);
+    release_version->version = std::string(version);
+    release_version->build_number =
+        std::to_string(LOWORD(file_info->dwFileVersionLS));
+    return kSuccess;
+  }
+
+  return -1;
+}
+
+std::string GetLocalAppDataPath() {
+  PWSTR path = nullptr;
+  HRESULT result = SHGetKnownFolderPath(FOLDERID_LocalAppData, 0, NULL, &path);
+  if (SUCCEEDED(result)) {
+    std::wstring widePath(path);
+    std::string localAppDataPath(widePath.begin(), widePath.end());
+    // The calling process is responsible for freeing this resource
+    // https://learn.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetknownfolderpath
+    CoTaskMemFree(path);
+    return localAppDataPath;
+  } else {
+    FML_LOG(ERROR) << "Failed to retrieve the local AppData directory.";
```

Comment:
I think it's more cannonical to return a failrue code (or even bool) and then fill an outparam, but this can work, just seems plausible that failure will only result in a log and continuing with an empty string (and thus failing in a different way later).

Reply:
Updated